pkg/sdk/testint/data_types_integration_test.go

```diff
@@ -0,0 +1,332 @@
+package testint
+
+import (
+	"fmt"
+	"slices"
+	"testing"
+
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/sdk/datatypes"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestInt_DataTypes(t *testing.T) {
+	client := testClient(t)
+	ctx := testContext(t)
+
+	incorrectBooleanDatatypes := []string{
+		"BOOLEAN()",
+		"BOOLEAN(1)",
+		"BOOL",
+	}
+	incorrectFloatDatatypes := []string{
+		"DOUBLE()",
+		"DOUBLE(1)",
+		"DOUBLE PRECISION(1)",
+	}
+	incorrectlyCorrectFloatDatatypes := []string{
+		"FLOAT()",
+		"FLOAT(20)",
+		"FLOAT4(20)",
+		"FLOAT8(20)",
+		"REAL(20)",
+	}
+	incorrectNumberDatatypes := []string{
+		"NUMBER()",
+		"NUMBER(x)",
+		"INT()",
+		"NUMBER(36, 5, 7)",
+	}
+	incorrectTextDatatypes := []string{
+		"VARCHAR()",
+		"VARCHAR(x)",
+		"VARCHAR(36, 5)",
+	}
+	// can't use slices.Concat on 1.21
+	vectorInnerTypesSynonyms := make([]string, 0)
+	vectorInnerTypesSynonyms = append(vectorInnerTypesSynonyms, datatypes.AllNumberDataTypes...)
+	vectorInnerTypesSynonyms = append(vectorInnerTypesSynonyms, datatypes.FloatDataTypeSynonyms...)
+	vectorInnerTypeSynonymsThatWork := []string{
+		"INTEGER",
+		"INT",
+		"FLOAT8",
+		"FLOAT4",
+		"FLOAT",
+	}
+
+	for _, c := range datatypes.ArrayDataTypeSynonyms {
+		t.Run(fmt.Sprintf("check behavior of array datatype: %s", c), func(t *testing.T) {
+			sql := fmt.Sprintf("SELECT []::%s", c)
+			_, err := client.QueryUnsafe(ctx, sql)
+			assert.NoError(t, err)
+
+			sql = fmt.Sprintf("SELECT []::%s(36)", c)
+			_, err = client.QueryUnsafe(ctx, sql)
+			assert.Error(t, err)
+		})
+	}
+
+	for _, c := range datatypes.BinaryDataTypeSynonyms {
+		t.Run(fmt.Sprintf("check behavior of binary datatype: %s", c), func(t *testing.T) {
+			sql := fmt.Sprintf("SELECT TO_BINARY('AB')::%s", c)
+			_, err := client.QueryUnsafe(ctx, sql)
+			assert.NoError(t, err)
+
+			sql = fmt.Sprintf("SELECT TO_BINARY('AB')::%s(36)", c)
+			_, err = client.QueryUnsafe(ctx, sql)
+			assert.NoError(t, err)
+
+			sql = fmt.Sprintf("SELECT TO_BINARY('AB')::%s(36, 2)", c)
+			_, err = client.QueryUnsafe(ctx, sql)
+			assert.Error(t, err)
```

Comment:
Can we assert what it really returns, just for the future reference? In other cases as well.

Reply:
Added